### ISSUE TITLE
Allow marking an individual local record as "resolved"

### ISSUE THREAD

**Issue by n1k0** [Kinto/kinto.js#414]
We don't currently have any API to mark a record as *resolved* after a conflict has been handled locally (typically using the `Collection#resolve()` method). It would probably make sense to schedule these for being individually force-published (`CLIENT_WINS` strategy) at next `sync()` call.

What do you think? /cc @leplatrem @Natim @almet @glasserc

**Comment by glasserc** [comment]
I don't like it. This really only works in the case where we have CLIENT_WINS, otherwise we could lose data.

I was able to reproduce it using Kinto/kinto-admin#54. At time t=t0, I created a record locally and synced it. At t=t1, I changed it in the browser. At t=t2, I changed it on the Kinto server using httpie. At t=t3, I tried to do another sync and got a conflict. I resolved the conflict and retried the sync and still got the conflict (as @n1k0 reported).

I think what we're seeing is more like a bug in Kinto.js. What happens is:

- The client initiates a sync by checking for any changes since time t0, which is when the last sync happened.
- The Kinto server reports that at t2, there was a change locally.
- When I resolve the change, it correctly updates the record's last_modified to be t2.
- However, when I do another sync, it begins by asking again for any changes since time t0.

The first thing I thought of was to make it save the new last modified on every conflict resolution, but of course that won't work because it can hide/cause us to "forget" other conflicts. I think the right fix here is that when the client does a sync, it should still check for changes since t0, but when it sees that one record was changed at t2 and that our local copy was "resolved" with the t2 version (according to its last_modified), it shouldn't consider it a conflict. It should still only update the collection's last_modified when no conflicts happen (or when the only conflicts have already been resolved).

**Comment by glasserc** [comment]
I tried to write the above up as a test, etc. but I broke other tests and I'm not confident that I wrote the test correctly anyhow. You can see it at https://github.com/glasserc/kinto.js/tree/conflicts.

**Comment by leplatrem** [comment]
I'm thinking out loud: when do we resolve a conflict, why don't we set the sync status to «updated»? It would be equivalent to having accepted the server version and applied changes on top of it.

Or not ? :)

**Comment by n1k0** [comment]
I hate you.

**Comment by n1k0** [comment]
@glasserc after discussing it with @leplatrem, we think you just nailed it on the head. Writing a test for this is no easy task, we'll help you.

**Comment by leplatrem** [comment]
Something like:
* In resolve() mark as `"synced"` if conflict is resolved with remote version. `"updated"` otherwise.
* Ignore incoming conflict `if (... && local.last_modified === remote.last_modified)` as you spotted
* Add them to `resolved: [...]` of sync result object

The `pushChanges()` will be called  (since `syncResult.ok == true`) and the new resolved version with be published to the server.

**Comment by leplatrem** [comment]
Note:
```
mai 06 08:01:56 <glasserc>	[04:20:46] leplatrem: J'ai cherché le pourquoi "resolved" n'a pas pris avec les tests, c'est parce qu'il manque un case dans pushChanges
mai 06 08:01:56 <glasserc>	[04:21:38] S'il y a "resolved" qui est "synced" mais strategy est MANUAL, on ne renvoie rien
```

**Comment by glasserc** [comment]
To elaborate, while I was working on #417, I tried at first to make `importChange` return a `resolved` status when it saw a conflict that had been previously resolved, but `pushChanges` didn't handle that correctly and this caused my test to break. I've opened #418 to try to explain exactly what I think the problem is.

**Comment by n1k0** [comment]
> * In resolve() mark as `"synced"` if conflict is resolved with remote version. `"updated"` otherwise.

That sounds reasonable, any reason this has not been done in this patch?

> * Add them to `resolved: [...]` of sync result object

Note that [docs for SyncResultObject.resolved](http://kintojs.readthedocs.io/en/latest/api/#the-synchronization-result-object) are:

> resolved: The list of conflicting records which have been successfully resolved according to the selected strategy (note that when using the default MANUAL strategy, this list is always empty).

**Comment by leplatrem** [comment]
Fixed in #417
